[WC-93] Accordion widget (web) - Add basic mode features

IIsaku opened the pull request:
## Checklist
- Contains unit tests ✅ 
- Contains breaking changes ❌
- Contains Atlas changes ❌
- Compatible with: MX 9️⃣

#### Web specific
- Contains e2e tests ❌
- Is accessible ❌
- Compatible with Studio ✅ 
- Cross-browser compatible ✅ 

#### Feature specific
- Comply with designs ❌
- Comply with PM's requirements ✅ 

## This PR contains
- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Other (describe)

## What is the purpose of this PR?
This PR adds the basic mode features for the new Accordion widget for web.

## Relevant changes
- Configure groups with a header & content.
- Lazy load the main content of a group.
- Configure the header of a group with text or widgets.
- Make header config props conditionally visible.
- Hide groups conditionally based on a boolean expression.
- Control whether the content of groups can collapse.
- Control the collapse behavior by allowing one or more groups to be expanded at the same time.
- Transform widget property groups into tabs for Studio.

## What should be covered while testing?
- Test the workings of the basic mode features.
- Pay special attention to group visibility in combination with collapsed group state.
- Test the lazy content loading feature on startup.
- Verify that group content remains in the DOM and is hidden when this group has been expanded at least once.

IIsaku committed:
chore(accordion-web): scaffold widget package

IIsaku committed:
feat(accordion-web): add group object list property

IIsaku committed:
chore(accordion-web): move editor config inside src dir

IIsaku committed:
chore(accordion-web): add collapse behaviour widget config

IIsaku committed:
feat(accordion-web): add group rendering

IIsaku committed:
chore(accordion-web): add test project

IIsaku committed:
chore(accordion-web): include fixed test project

IIsaku committed:
feat(accordion-web): add content widget config property

IIsaku committed:
feat(accordion-web): change collapse behavior widget config property

- Change enumeration value description.

IIsaku committed:
feat(accordion-web): render groups

- Render the content and header of all configured groups

IIsaku committed:
chore(accordion-web): bump test project ref

IIsaku committed:
feat(accordion-web): toggle content visibility

- Groups are closed on first render.
- Content node will be mounted on first expansion and will remain mounted afterwards.

IIsaku committed:
feat(accordion-web): implement collapsible property

- Show all content when collapsible is false.
- Header isn't clickable when collapsible is false.

IIsaku committed:
feat(accordion-web): implement collapsed prop for AccordionGroup

- Create a prop with which you can influence the collapsed state of the group

IIsaku committed:
refactor(accordion-web): collapsed prop implementation for AccordionGroup

IIsaku committed:
feat(accordion-web): single expanded accordion group

IIsaku committed:
feat(accordion-web): add visibility property for each group

- With an boolean expression widget configuration
property the visibility of a group can be influenced.
- The collapsed state stays in tact and keeps being
updated accordingly.
- If the boolean expression cannot be evaluated or the
header text can't be retrieved, the accordion isn't
rendered.

IIsaku committed:
feat(accordion-web): require one group to be configured

IIsaku committed:
fix(accordion-web): accordionGroup mounting logic & state/prop syncing

IIsaku committed:
fix(accordion-web): multiple expanded accordion groups

- In single expanded mode, accordion groups are never
open at the same time due to moving all state to the
component inside src/components/Accordion.tsx.

IIsaku committed:
refactor(accordion-web): accordion group state management to useReducer

IIsaku committed:
feat(accordion-web): allow multiple expanded accordion groups

IIsaku committed:
chore(accordion-web): adjust changelog description

IIsaku committed:
chore(accordion-web): delete spec tsconfig

IIsaku committed:
chore(accordion-web): update test project ref

IIsaku committed:
chore(accordion-web): remove cross-env dependency

IIsaku committed:
chore(accordion-web): setup unit test config

IIsaku committed:
fix(accordion-web): not collapsible mode

diego-antonelli reviewed (changes requested):
Good PR

IIsaku committed:
refactor(accordion-web): remove collapsible prop

- Remove collapsible prop for AccordionGroup.

IIsaku committed:
test(accordion-web): add AccordionGroup unit tests

IIsaku committed:
refactor(accordion-web): function AccordionGroupsReducer

IIsaku committed:
test(accordion-web): accordion groups reducer

IIsaku committed:
refactor(accordion-web): move accordion groups reducer

IIsaku committed:
refactor(accordion-web): rename accordion groups reducer unit test file

IIsaku committed:
test(accordion-web): unit test accordion component

IIsaku commented on packages/pluggableWidgets/accordion-web/src/Accordion.tsx:
```diff
@@ -0,0 +1,56 @@
+    const accordionGroups: AccGroup[] | undefined = useMemo(() => {
+        const result = [];
+
+        for (const group of groups) {
```
My first implementation was with a `map`, but I need to prevent rendering (by returning `undefined`) as soon as some group value isn't available yet.

diego-antonelli replied:
As we talked, you are gonna check with `array.some` and change it to map :)

IIsaku replied:
Changed it.

IIsaku commented on packages/pluggableWidgets/accordion-web/src/Accordion.xml:
```diff
@@ -0,0 +1,66 @@
+                                <caption>Header</caption>
+                                <description/>
+                                <enumerationValues>
+                                    <enumerationValue key="text">Text</enumerationValue>
```
The requirements paper doc says "Text". I will propose your suggestion to Chris.

diego-antonelli replied:
In DG We use Dynamic Text and Custom content

IIsaku replied:
Waiting for a response from PMs

IIsaku committed:
refactor(accordion-web): existence check dispatch before executing

IIsaku committed:
test(accordion-web): fix unit test

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -0,0 +1,51 @@
+
+    const toggleContentVisibility = useCallback(() => {
+        if (group.collapsed) {
+            dispatch!({ type: "expand", group });
```
I agree. Changed it.

IIsaku commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -0,0 +1,88 @@
+            return state;
+        }
+
+        if (action.type === "sync" || singleExpandedGroup) {
```
I changed the implementation in the meanwhile. Do you think it's still useful now?

IIsaku committed:
refactor(accordion-web): remove destructuring in glue component

IIsaku committed:
chore(accordion-web): improve widget description

IIsaku commented on packages/pluggableWidgets/accordion-web/package.json:
```diff
@@ -0,0 +1,35 @@
+  "name": "accordion-web",
+  "widgetName": "Accordion",
+  "version": "1.0.0",
+  "description": "A Mendix pluggable widget to toggle the display of sections of content.",
```
True, I pushed an improvement. What do you think?

diego-antonelli replied:
better :)

IIsaku commented on packages/pluggableWidgets/accordion-web/src/Accordion.tsx:
```diff
@@ -0,0 +1,56 @@
+import { AccGroup } from "./components/AccordionGroup";
+
+export function Accordion(props: AccordionContainerProps): ReactElement | null {
+    const { class: className, name, style, tabIndex, groups, collapsible, collapseBehavior } = props;
```
I agree and changed it 🙂

diego-antonelli commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -0,0 +1,42 @@
+        groups.map(group => (collapsible ? { ...group, collapsed: true } : { ...group, collapsed: false }))
+    );
+
+    if (groups !== previousGroupsPropValue.current) {
```
Dont we need to compare using deep-equal here?

IIsaku replied:
No the array will always be different when there is a change in one of the group values, because of the `useMemo` implementation in `src/Accordion.tsx`

IIsaku committed:
refactor(accordion-web): accordion components

- Refactor accordion group iterating in src/Accordion.tsx.
- Remove destructuring in src/components/Accordion.tsx.

Keraito commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -0,0 +1,40 @@
+
+    const [accordionGroups, accordionGroupsDispatch] = useReducer(
+        getAccordionGroupsReducer(props.collapsible ? (props.singleExpandedGroup ? "single" : "multiple") : "all"), // the accordion group reducer function doesn't need to change during the lifetime of this component, since the singleExpandedGroup won't change.
+        props.groups.map(group => (props.collapsible ? { ...group, collapsed: true } : { ...group, collapsed: false }))
```
```suggestion
        props.groups.map(group => ({ ...group, collapsed: props.collapsible }))
```

Keraito commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -0,0 +1,40 @@
+        props.groups.map(group => (props.collapsible ? { ...group, collapsed: true } : { ...group, collapsed: false }))
+    );
+
+    if (props.groups !== previousGroupsPropValue.current) {
```
Question: wouldn't it better to make this a check onto the content of the arrays rather than rely on the `useMemo` from the parent component? Right now this will trigger anyways if let say an update happens with the same content through dynamic values. In that sense I feel like doing a content comparison is more reilable

IIsaku replied:
As discussed we created a separate story on the backlog for this.

Keraito commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -0,0 +1,50 @@
+export default function AccordionGroup(props: AccordionGroupProps): ReactElement | null {
+    const { group } = props;
+
+    const dispatch = useContext(AccordionGroupsDispatch);
```
What is the reason we chose to use a Context for an one level deep callback? I feel like opening up an `onToggle: (isCollapsed: boolean) => void` API prop would suffice in case. Am I missing something?

Keraito replied:
That way you can also avoid the optional chaining on the dispatch in your `toggleContentVisibility` and handle it in the parent

IIsaku replied:
Converted to a callback prop. I like your idea to create a nice `onToggle` callback prop that extracts all logic, but for that we would need to memoize functions inside the map function, because we cannot use useCallback inside map functions. This would require work, which I want to prevent for this PR. I would add it as a separate task to the board to improve it. What do you think?

Keraito commented on packages/pluggableWidgets/accordion-web/src/utils/AccordionGroupStateReducer.ts:
```diff
@@ -0,0 +1,41 @@
+            });
+        }
+
+        const newState = [...state];
```
This can just be `const newState = state` or not?

IIsaku replied:
The state needs to be a new array in order to be picked up ([React docs](https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-dispatch)):

> If you return the same value from a Reducer Hook as the current state, React will bail out without rendering the children or firing effects.

diego-antonelli replied:
Nope, if we change the reference it doesnt trigger re-render.

Keraito commented on packages/pluggableWidgets/accordion-web/src/utils/AccordionGroupStateReducer.ts:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        const newState = [...state];
+        const groupIndex = newState.findIndex(group => group === action.group);
```
Can we just ignore if `groupIndex` is -1?

IIsaku replied:
Changed it. Also included a unit test.

Keraito commented on packages/pluggableWidgets/accordion-web/src/utils/AccordionGroupStateReducer.ts:
```diff
@@ -0,0 +1,41 @@
+        const groupIndex = newState.findIndex(group => group === action.group);
+
+        if (action.type === "expand") {
+            newState[groupIndex] = { ...action.group, collapsed: false };
```
If you are structuring your reducer function like this (with if statements instead of switch case), I would suggest the following: `newState[groupIndex] = { ...action.group, collapsed: action.type !== "expand" };`

IIsaku replied:
Done.

Keraito commented on packages/pluggableWidgets/accordion-web/src/utils/AccordionGroupStateReducer.ts:
```diff
@@ -0,0 +1,41 @@
+                let collapsed;
+
+                if (action.type === "sync") {
+                    collapsed = state[index].collapsed; // The previous collapsed state of the group at the same index can be used, because the order of groups remains the same throughout the lifetime of the widget.
```
- So if I'm correct, the `sync` action just syncs in the incoming `action.groups` AccGroup items but keeps its previous `collapsed` state?
- If so, why is the behaviour not that the new group just overrides everything?
- Can this widget be used with dynamic values? If so, does the comment still hold?

IIsaku replied:
- Correct.
- You don't want to override everything, because that would reinit the collapsed state on every data change.
- The ReactNode get updated when there is a change in the contained widgets. Also the visibility prop is dynamic. Therefore, this comment doesn't hold.

Keraito commented on packages/pluggableWidgets/accordion-web/src/components/Accordion.tsx:
```diff
@@ -0,0 +1,40 @@
+        accordionGroupsDispatch({ type: "sync", groups: props.groups });
+    }
+
+    const renderedGroups = useMemo(() => {
+        return accordionGroups.map((group, index) => <AccordionGroup key={index} group={group} />);
+    }, [accordionGroups]);
```
This feels a bit overkill, did you come across issues with this? Because I would assume that the `key` would be enough to prevent this from re-rendering.

IIsaku replied:
Removed the useMemo.

Keraito commented on packages/pluggableWidgets/accordion-web/src/Accordion.tsx:
```diff
@@ -0,0 +1,36 @@
+export function Accordion(props: AccordionContainerProps): ReactElement | null {
+    const accordionGroups: AccGroup[] | undefined = useMemo(() => {
+        if (props.groups.some(group => group.visible.value === undefined || group.headerText.value === undefined)) {
+            return undefined;
```
Is this per design? Because It feels weird that if one according group is undefined, we just stop rendering all together. Wouldn't just filtering it out make more sense? Or am I missing something?

IIsaku replied:
I think the first time the widget becomes visible it should always be fully loaded. That's how it's done in many other widgets as well. Anyhow, I will reach out to Chris to verify whether this is the desired behaviour. If he wants to change the behaviour, I will create a separate story for it. So you may resolve this conversation 😬

Keraito replied:
Correct me if I'm wrong, but the conditional indicates to me that "if any group visibility is not defined or has no headertext, then we decide to not show any group", which doesn't sound too "it should always be fully loaded". Or am I missing something?

IIsaku replied:
I am not following you fully here. Both values will be `undefined` in the case that the data isn't available yet, which can only happen when mounting the widget. Once all data has been loaded and calculated in the client, we want to render. After that, the values changes will be applied immediately, because they will never be `undefined` anymore.

Keraito replied:
Discussed this with Isa. We need some more information from Jelmer whether there are realistic cases where either of these values can become `undefined` during runtime (when the dynamic value turns unavailable). Because if so, the accordion widget will just stop rendering which I assume is not what we want. I will approve meanwhile but it is an improvement if it is not a super edge case

tamelbrun replied:
If the accordion widget is inside a data container that is not guaranteed to always have an object (for example, if it's listening to a data grid which doesn't always have a selected object), then the visibility expressions and header texts can become unavailable, but only if they depend on the unavailable object. Standard behaviour in Mendix is to be invisible if the visibility expression cannot be evaluated, but that could be different for each individual group. I think it would make more sense to simply filter on the truthiness of the visibility expression when deciding which groups to render (and render nothing only if no groups are visible at all).

I'm not sure what should happen if the visibility expression evaluates to `true` and the header text is unavailable, though. It currently renders no groups at all if any header text is unavailable, even if the `headerRenderMode` is not text. My suggestion would be to always render the group if the visibility expression says that it's visible, as it's always possible to configure a fallback text for the header if it's unavailable (and in non-text mode, you should probably render the `headerContent` anyway).

IIsaku replied:
Created a [Jira issue](https://mendix.atlassian.net/browse/WC-508) to fix this.

Keraito commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -0,0 +1,50 @@
+    return (
+        <section>
+            <header onClick={dispatch ? toggleContentVisibility : undefined}>{group.header}</header>
+            <div style={{ display: dispatch && group.collapsed ? "none" : undefined }}>
```
Is there a reason this is done through inline styling render than conditional rendering on the div through JavaScript? Animations?

IIsaku replied:
I will move this to functional styles in the next story. :)

Keraito commented on packages/pluggableWidgets/accordion-web/src/utils/AccordionGroupStateReducer.ts:
```diff
@@ -0,0 +1,41 @@
+export function getAccordionGroupsReducer(
+    expand: "single" | "multiple" | "all"
+): (state: AccGroup[], action: AccordionGroupsReducerAction) => AccGroup[] {
+    return (state: AccGroup[], action: AccordionGroupsReducerAction): AccGroup[] => {
```
Nit: For future reference: a switch case is often times more readable since it forces you to split the cases. Now a few cases are intertwined which makes it slightly hard to follow.

IIsaku replied:
I agree. I will try that next time :)

Keraito reviewed (changes requested):
Got a few questions and not sure some of the construction decisions were really necessary. Requesting changes is mainly for the usage of context since I think it's quite overkill and introduced quite a bit of maintenance burden + relying on `useMemo` to check for equality rather than content.

IIsaku committed:
refactor(accordion-web): simplify accordion group state init

Co-authored-by: Chak Shun Yu <chakshunyu@gmail.com>

IIsaku committed:
refactor(accordion-web): accordion component

IIsaku committed:
feat(accordion-web): improve accordion group reducer

- If the group cannot be found in state anymore, we return the previous state.

IIsaku committed:
refactor(accordion-web): convert useContext to callback prop

IIsaku committed:
refactor(accordion-web): reducer

Keraito commented on packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx:
```diff
@@ -0,0 +1,50 @@
+        }
+    }, [group, setDivContentMounted]);
+
+    const toggleContentVisibility = useCallback(() => {
+        if (group.collapsed) {
+            accordionGroupsDispatch!({ type: "expand", group });
+        } else {
+            accordionGroupsDispatch!({ type: "collapse", group });
+        }
+    }, [group, accordionGroupsDispatch]);
```
Regarding the `onToggle`: I would just remove this callback alltogether and pull it up to `Accordion.tsx`. Right now this component is not optimized for re-render anyways because `group` is an object and not memoized (the array is, but not this object). I would suggest just breaking up all the properties from the `AccGroup` object and make them individual APIs, so:
```
export interface AccordionGroupProps {
    header: ReactNode;
    content: ReactNode;
    collapsed?: boolean;
    visible: boolean;
    onToggle: (isCollapsed: boolean) => void;
}
```

I feel like from a React design perspective this also makes more sense for its composability. Then in the parent you can just do `onToggle={(isCollapsed) => accordionGroupsDispatch({ type: isCollapsed ? "expand" : "collapse", group })}`. I don't think performance or memoization should really be a concern here since the tree doesn't go too deep after this. If you really wanted you could memoize per group, but that might be quite overkill.

IIsaku replied:
I agree with all of the above. I will make sure to include this in a story on our backlog to improve.

IIsaku replied:
[Jira issue](https://mendix.atlassian.net/browse/WC-506)

IIsaku committed:
chore(accordion-web): disable e2e test script call
